pkg/manifest/ostree_deployment.go

```diff
@@ -193,51 +188,58 @@ func (p *OSTreeDeployment) serialize() osbuild.Pipeline {
 		)
 	}
 
-	pipeline.AddStage(osbuild.NewOSTreePullStage(
-		&osbuild.OSTreePullStageOptions{Repo: repoPath, Remote: p.Remote.Name},
-		osbuild.NewOstreePullStageInputs("org.osbuild.source", commit.Checksum, commit.Ref),
-	))
-
-	pipeline.AddStage(osbuild.NewOSTreeDeployStage(
-		&osbuild.OSTreeDeployStageOptions{
-			OsName: p.osName,
-			Ref:    commit.Ref,
-			Remote: p.Remote.Name,
-			Mounts: []string{"/boot", "/boot/efi"},
-			Rootfs: osbuild.Rootfs{
-				Label: "root",
+	var ref string
+	if p.ostreeSpec != nil {
+		commit := *p.ostreeSpec
+		ref = commit.Ref
+		pipeline.AddStage(osbuild.NewOSTreePullStage(
+			&osbuild.OSTreePullStageOptions{Repo: repoPath, Remote: p.Remote.Name},
+			osbuild.NewOstreePullStageInputs("org.osbuild.source", commit.Checksum, ref),
+		))
+
+		pipeline.AddStage(osbuild.NewOSTreeDeployStage(
+			&osbuild.OSTreeDeployStageOptions{
+				OsName: p.osName,
+				Ref:    ref,
+				Remote: p.Remote.Name,
+				Mounts: []string{"/boot", "/boot/efi"},
+				Rootfs: osbuild.Rootfs{
+					Label: "root",
+				},
+				KernelOpts: kernelOpts,
 			},
-			KernelOpts: kernelOpts,
-		},
-	))
+		))
 
-	remoteURL := p.Remote.URL
-	if remoteURL == "" {
-		// if the remote URL for the image is not specified, use the source commit URL
-		remoteURL = commit.URL
-	}
-	pipeline.AddStage(osbuild.NewOSTreeRemotesStage(
-		&osbuild.OSTreeRemotesStageOptions{
-			Repo: "/ostree/repo",
-			Remotes: []osbuild.OSTreeRemote{
-				{
-					Name:        p.Remote.Name,
-					URL:         remoteURL,
-					ContentURL:  p.Remote.ContentURL,
-					GPGKeyPaths: p.Remote.GPGKeyPaths,
+		remoteURL := p.Remote.URL
+		if remoteURL == "" {
+			// if the remote URL for the image is not specified, use the source commit URL
+			remoteURL = commit.URL
+		}
+		pipeline.AddStage(osbuild.NewOSTreeRemotesStage(
+			&osbuild.OSTreeRemotesStageOptions{
+				Repo: "/ostree/repo",
+				Remotes: []osbuild.OSTreeRemote{
+					{
+						Name:        p.Remote.Name,
+						URL:         remoteURL,
+						ContentURL:  p.Remote.ContentURL,
+						GPGKeyPaths: p.Remote.GPGKeyPaths,
+					},
 				},
 			},
-		},
-	))
+		))
```

Comment:
Is this stage always added? IIUC it's not always needed. For example, in Fedora we deliver remote configuration via the `fedora-repos-ostree` RPM:

```
[core@rpi4 ~]$ rpm -ql fedora-repos-ostree
/etc/ostree/remotes.d
/etc/ostree/remotes.d/fedora-compose.conf
/etc/ostree/remotes.d/fedora.conf
```

- [source](https://src.fedoraproject.org/rpms/fedora-repos/blob/rawhide/f/fedora.conf)

Reply:
We used to configure this before the package was created.  I have [a separate PR](https://github.com/osbuild/images/pull/95) that's been open for a while but configuring the remote through the package has been causing issues with our tests in osbuild-composer (https://github.com/osbuild/osbuild-composer/pull/3635).

I've been looking into it on and off and have no idea what the issue is, but I'd love to get your input on it.

Reply:
looks like https://github.com/ostreedev/ostree/discussions/3079#discussioncomment-7339766 seems to have figured it all out?

Reply:
It seems to be something else.  I should write it up somewhere else but here's how the problem manifests:
- Build an image with the ostree remote stage (like we do now).
- Boot it
- Remove the remote
- Add a new remote with the same name, but with a different URL
- Add an upgrade commit to the location of the new remote
- Run `rpm-ostree upgrade`

This works.  The update gets pulled and all is well.

If I build an image without the ostree remote stage but with the packaged remote configuration and perform the same steps, then `rpm-ostree upgrade` doesn't do anything.  I can `ostree remote log` and `ostree remote summary` etc, I can see the upgrade commit in the repo.  I can even `rpm-ostree rebase` onto it.  But `rpm-ostree upgrade` refuses to pick it up.

Reply:
dumb question.. there is a new commit in the history for that ref to upgrade to?

Reply:
Yes, I create the repo, serve it over a web server, and pull an upgrade commit into it in both scenarios.

I'd love to show you the whole thing.  I'll tidy up my repro scripts and put them somewhere tomorrow as well, show them around to anyone who might have some idea about what's going on.

Reply:
Solved this issue after 5 mins of showing Dusty the issue.  I wasn't defining a remote name when deploying.  Fixed separately.  Thanks @dustymabe!!